### ISSUE TITLE
How to implement swagger express UI in https domain?

### ISSUE THREAD

**Issue by turivishal** [scottie1984/swagger-ui-express#237]
I am using below code to generate swagger UI in node.js and express:
```
const swaggerUi = require('swagger-ui-express');
app.use('/', swaggerUi.serve, swaggerUi.setup(
    require('./swagger.json'),
    {
        explorer: true,
        swaggerOptions: {
            displayRequestDuration: true,
            docExpansion: "none",
            filter: false,
            showExtensions: true,
            showCommonExtensions: true,
            displayOperationId: true,
            urls: [] // list of urls
        }
    }
));
```

I am using Docker to deploy node.js project:
```
FROM node:13-alpine
COPY ./package.json ./
RUN apk add git
RUN npm install --production
COPY . .
EXPOSE // some valid port
CMD ["node","app.js"]
```

**Working on http and with port** `http://mydomain.com:<port>`,

**Not working on https** `https://mydomain.com`, it show white screen shows error in browser's console:

Error 1: 
```
Failed to load resource: the server responded with a status of 502 (Bad Gateway)
swagger-ui-bundle.js:1 Failed to load resource: the server responded with a status of 502 (Bad Gateway)
swagger-ui-standalone-preset.js:1 Failed to load resource: the server responded with a status of 502 (Bad Gateway)
swagger-ui-init.js:1 Failed to load resource: the server responded with a status of 502 (Bad Gateway)
favicon-32x32.png:1 GET https://mydomain.com/favicon-32x32.png 502 (Bad Gateway)
```
Error 2:
```
Uncaught SyntaxError: Unexpected token '<'
swagger-ui-standalone-preset.js:3 Uncaught SyntaxError: Unexpected token '<'
swagger-ui-init.js:3 Uncaught SyntaxError: Unexpected token '<'
```

**Is there any process or configuration to work in SSL domain?**

**Comment by 01speed1** [comment]
I have a similar issue i dont know why with a HTTP request the CSS and JS call an HTTPS request, what can i do ? 

![Screen Shot 2021-03-20 at 15 36 35](https://user-images.githubusercontent.com/7319582/111884830-12e28e00-8992-11eb-9a66-04aabd0bdce8.png)

**Comment by arunvishnun** [comment]
@01speed1 Were you able to figure out a solution for this. Facing same issue!!

**Comment by marcobertell** [comment]
> I have a similar issue i dont know why with a HTTP request the CSS and JS call an HTTPS request, what can i do ?
> 
> ![Screen Shot 2021-03-20 at 15 36 35](https://user-images.githubusercontent.com/7319582/111884830-12e28e00-8992-11eb-9a66-04aabd0bdce8.png)

i have the same issue.

**Comment by mostafaebrahimi** [comment]
> I have a similar issue i dont know why with a HTTP request the CSS and JS call an HTTPS request, what can i do ?
> 
> ![Screen Shot 2021-03-20 at 15 36 35](https://user-images.githubusercontent.com/7319582/111884830-12e28e00-8992-11eb-9a66-04aabd0bdce8.png)

Same Issue! (nginx as reverse-proxy)

**Comment by mostafaebrahimi** [comment]
@turivishal Did you found any solution?
I checked a lot of things also `swagger-ui-express` code but didn't find anything

**Comment by mostafaebrahimi** [comment]
@turivishal seems this problem is not related to `swagger-ui-express` and it's related to your express configuration

**Comment by turivishal** [comment]
@mostafaebrahimi I am not sure, is there any other configuration for HTTPS domain? It's working in HTTP domain.

**Comment by mostafaebrahimi** [comment]
@turivishal paste your express config here.

**Comment by inventorypro1** [comment]
I have the same issue. I can define it as the following:

Relevant versions: 
 - Node: 14.16.0
 - Express: 4.17.1
 - swagger-ui-express: 4.1.6
 - swagger-jsdoc: 6.1.0

I have a swagger-docs.js file in my project which populates the relevant routes in the following manner:
```
function initSwaggerRoutes(app) {
    if (args['swagger.enabled'] !== undefined && args['swagger.enabled'] === true) {
        app.use('/swagger-ui.html', swaggerUi.serve, swaggerUi.setup(swaggerSpec, {}, { validatorUrl: null }));
        app.use('/swagger-resources', swaggerResources);
        app.use('/api-docs', apiDocs);
    }
    else {
        log.info('Disabling swagger');
    }
}
```
I get the following error message when running on a live environment:
<img width="556" alt="Screen Shot 2021-08-18 at 10 12 37" src="https://user-images.githubusercontent.com/77077215/129871833-5ee2ebdc-a4a6-49f4-9cc5-5d9f6d59301b.png">
This results in a blank screen displaying when I navigate to `http://<serverIP>:<serverPort>/swagger-ui.html`

The issue is occurring as for some reason the request is being redirected to a HTTPS endpoint to retrieve the relevant documents, even though I am requesting a http endpoint.

This issue **does not** occur on `swagger-ui-express` version 4.1.3 or lower. I have confirmed this fact as 9 of my microservices use version 4.1.3 and does not have the issue (they all have identical configuration other than the version used). 5 of my microservices using version 4.1.6 has the issue as described.

**Comment by marcdefaria** [comment]
I have the same issue. I can define it as the following:

Relevant versions: 
 - Node: 14.16.0
 - Express: 4.17.1
 - swagger-ui-express: 4.1.6
 - swagger-jsdoc: 6.1.0

(Using docker for deployment)

I have a swagger-docs.js file in my project which populates the relevant routes in the following manner:
```
function initSwaggerRoutes(app) {
    if (args['swagger.enabled'] !== undefined && args['swagger.enabled'] === true) {
        app.use('/swagger-ui.html', swaggerUi.serve, swaggerUi.setup(swaggerSpec, {}, { validatorUrl: null }));
        app.use('/swagger-resources', swaggerResources);
        app.use('/api-docs', apiDocs);
    }
    else {
        log.info('Disabling swagger');
    }
}
```
I get the following error message when running on a live environment:
<img width="556" alt="Screen Shot 2021-08-18 at 10 12 37" src="https://user-images.githubusercontent.com/20932787/129872842-292bb669-1e67-4943-8742-d1bf08130b5e.png">

The issue is occurring as for some reason the request is being redirected to a HTTPS endpoint to retrieve the relevant documents, even though I am requesting a http endpoint.

This issue **does not** occur on `swagger-ui-express` version 4.1.3 or lower. I have confirmed this fact as 9 of my microservices use version 4.1.3 and does not have the issue (they all have identical configuration other than the swagger-ui-express version used). 5 of my microservices using version 4.1.6 has the issue as described.

**Comment by marcdefaria** [comment]
I just downpinned the version of swagger-ui-express for one of the services that had the above issue to use 4.1.3 (instead of 4.1.6) and I have the same issue. Seems like its perhaps something incorrect with my configuration that is causing the request to be redirected to https when retrieving the resources of the swagger-ui.html page:
<img width="534" alt="Screen Shot 2021-08-18 at 11 19 26" src="https://user-images.githubusercontent.com/20932787/129881921-5cb750aa-666d-4b16-bbcb-16425318dc83.png">

**Comment by rjbobeles** [comment]
Downgrading jsdoc to 6.0.1 from 6.1.0 solved the issue for me.

```
 "swagger-jsdoc": "^6.0.1",
 "swagger-ui-express": "^4.1.6",
```

**Comment by rjbobeles** [comment]
Downgrading JSDoc to 6.0.1 did the trick for me.

```
 "swagger-jsdoc": "^6.0.1",
 "swagger-ui-express": "^4.1.6",
```

**Comment by rjbobeles** [comment]
Downgrading JSDoc from 6.1.0 to 6.0.1 did the trick in my case.

```
 "swagger-jsdoc": "^6.0.1",
 "swagger-ui-express": "^4.1.6",
```

**Comment by marcdefaria** [comment]
Interestingly enough, the issue I described was completely unrelated to swagger-ui-express and rather had to do with the fact that I updated helmet from version 3 to version 4.

Helmet is a library that adds response headers and content-security-policies to your API responses.

In the update from helmet version 3 to 4, a default value was set for the content-security-policy header which included a CSP header of "upgrade-insecure-requests". This caused the requests from swagger to fetch the style files to redirect from http to https.

Swagger style files were being served over http, hence the ERR_SSL_PROTOCOL_ERROR when requesting them over https.

I resolved the issue by adding the following logic when applying helmet to my application;
```
const cspDefaults = helmet.contentSecurityPolicy.getDefaultDirectives();
delete cspDefaults['upgrade-insecure-requests'];

app.use(helmet({
    contentSecurityPolicy: { directives: cspDefaults }
}));
```

This may be a bit TMI, but I struggled with this for a couple of days and thought I'd put the solution here incase anyone is having the same issue.

**Comment by avvvet** [comment]
@marcdefaria thanks, works for me.

**Comment by krejko** [comment]
> Interestingly enough, the issue I described was completely unrelated to swagger-ui-express and rather had to do with the fact that I updated helmet from version 3 to version 4.
> 
> Helmet is a library that adds response headers and content-security-policies to your API responses.
> 
> In the update from helmet version 3 to 4, a default value was set for the content-security-policy header which included a CSP header of "upgrade-insecure-requests". This caused the requests from swagger to fetch the style files to redirect from http to https.
> 
> Swagger style files were being served over http, hence the ERR_SSL_PROTOCOL_ERROR when requesting them over https.
> 
> I resolved the issue by adding the following logic when applying helmet to my application;
> 
> ```
> const cspDefaults = helmet.contentSecurityPolicy.getDefaultDirectives();
> delete cspDefaults['upgrade-insecure-requests'];
> 
> app.use(helmet({
>     contentSecurityPolicy: { directives: cspDefaults }
> }));
> ```
> 
> This may be a bit TMI, but I struggled with this for a couple of days and thought I'd put the solution here incase anyone is having the same issue.



> Interestingly enough, the issue I described was completely unrelated to swagger-ui-express and rather had to do with the fact that I updated helmet from version 3 to version 4.

I am still encountering this same issue. I recently started a project using the [typescript-express-starter template](https://www.npmjs.com/package/typescript-express-starter) which already included default configurations for Helmet 5.1.1. and swagger-ui-express 4.5.0.  

I typically do development using vs-code server to centralize my repos on one machine. When I hit swagger using localhost on server itself, everything loads fine. However, whenever access swagger from a non-localhost address (such as direct IP or using the computer.local address) I get a white screen with many ERR_SSL_PROTOCOL_ERROR in the console. This is both from the server itself and when accessing from other machines. I have tried both altering helmet to use the contentSecurityPolicy and have removed its initialization entirely, but neither seem to have any effect. 

Are there any other workarounds I could try for this issue?

**Comment by krejko** [comment]
> I have tried both altering helmet to use the contentSecurityPolicy and have removed its initialization entirely, but neither seem to have any effect.

Of course as soon as I post this, swagger begins working without helmet initialized. This will suffice for now, but would still be curious if there are other workarounds that let me continue to use helmet.
